testing/endtoend/evaluators/operations.go

```diff
@@ -312,47 +331,93 @@ func proposeVoluntaryExit(ec *e2etypes.EvaluationContext, conns ...*grpc.ClientC
-
-	voluntaryExit := &ethpb.VoluntaryExit{
-		Epoch:          chainHead.HeadEpoch,
-		ValidatorIndex: exitedIndex,
+	versionedMarshaler, err := detect.FromState(stObj.Encoded)
+	if err != nil {
+		return errors.Wrap(err, "could not get state marshaler")
 	}
-	req := &ethpb.DomainRequest{
-		Epoch:  chainHead.HeadEpoch,
-		Domain: params.BeaconConfig().DomainVoluntaryExit[:],
+	st, err := versionedMarshaler.UnmarshalBeaconState(stObj.Encoded)
+	if err != nil {
+		return errors.Wrap(err, "could not get state")
 	}
-	domain, err := valClient.DomainData(ctx, req)
+	execIndices := []int{}
+	err = st.ReadFromEveryValidator(func(idx int, val state.ReadOnlyValidator) error {
+		if val.WithdrawalCredentials()[0] == params.BeaconConfig().ETH1AddressWithdrawalPrefixByte {
+			execIndices = append(execIndices, idx)
+		}
+		return nil
+	})
 	if err != nil {
 		return err
 	}
-	signingData, err := signing.ComputeSigningRoot(voluntaryExit, domain.SignatureDomain)
+	if len(execIndices) > numOfExits {
+		execIndices = execIndices[:numOfExits]
+	}
+
+	deposits, privKeys, err := util.DeterministicDepositsAndKeys(params.BeaconConfig().MinGenesisActiveValidatorCount)
 	if err != nil {
 		return err
 	}
-	signature := privKeys[exitedIndex].Sign(signingData[:])
-	signedExit := &ethpb.SignedVoluntaryExit{
-		Exit:      voluntaryExit,
-		Signature: signature.Marshal(),
+
+	var sendExit = func(exitedIndex primitives.ValidatorIndex) error {
+		voluntaryExit := &ethpb.VoluntaryExit{
+			Epoch:          chainHead.HeadEpoch,
+			ValidatorIndex: exitedIndex,
+		}
+		req := &ethpb.DomainRequest{
+			Epoch:  chainHead.HeadEpoch,
+			Domain: params.BeaconConfig().DomainVoluntaryExit[:],
+		}
+		domain, err := valClient.DomainData(ctx, req)
+		if err != nil {
+			return err
+		}
+		signingData, err := signing.ComputeSigningRoot(voluntaryExit, domain.SignatureDomain)
+		if err != nil {
+			return err
+		}
+		signature := privKeys[exitedIndex].Sign(signingData[:])
+		signedExit := &ethpb.SignedVoluntaryExit{
+			Exit:      voluntaryExit,
+			Signature: signature.Marshal(),
+		}
+
+		if _, err = valClient.ProposeExit(ctx, signedExit); err != nil {
+			return errors.Wrap(err, "could not propose exit")
+		}
+		pubk := bytesutil.ToBytes48(deposits[exitedIndex].Data.PublicKey)
+		ec.ExitedVals[pubk] = true
+		return nil
+	}
+
+	// Send exits for keys which already contain execution credentials.
+	for _, idx := range execIndices {
+		if err := sendExit(primitives.ValidatorIndex(idx)); err != nil {
+			return err
+		}
 	}
 
-	if _, err = valClient.ProposeExit(ctx, signedExit); err != nil {
-		return errors.Wrap(err, "could not propose exit")
+	// Send an exit for a non-exited validator.
+	for i := 0; i < numOfExits; {
+		randIndex := primitives.ValidatorIndex(rand.Uint64() % params.BeaconConfig().MinGenesisActiveValidatorCount)
+		if ec.ExitedVals[bytesutil.ToBytes48(privKeys[randIndex].PublicKey().Marshal())] {
+			continue
```

Comment:
Is there a possibility that the loop never terminates?

Reply:
The only way that happens is if we have set up our config in such a way where we have exited almost all keys from our validator set. If that were to happen, the network would start to behave wierdly anyway.